src/core/lv2/Lv2Options.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Lv2Options.cpp - Lv2Options implementation
+ *
+ * Copyright (c) 2020-2020 Johannes Lorenz <jlsf2013$users.sourceforge.net, $=@>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Lv2Options.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include <QtGlobal>
+
+
+std::set<LV2_URID> Lv2Options::s_supportedOptions;
+
+
+
+
+bool Lv2Options::isOptionSupported(LV2_URID key)
+{
+	return s_supportedOptions.find(key) != s_supportedOptions.end();
+}
+
+
+
+
+void Lv2Options::supportOption(LV2_URID key)
+{
+	Q_ASSERT(s_supportedOptions.find(key) == s_supportedOptions.end());
+	s_supportedOptions.insert(key);
+}
+
+
+
+
+void Lv2Options::createOptionVectors()
+{
+	// create vector of options
+	for(LV2_URID urid : s_supportedOptions)
+	{
+		auto itr = m_optionByUrid.find(urid);
+		Q_ASSERT(itr != m_optionByUrid.end());
+		m_options.push_back(itr->second);
+	}
+	LV2_Options_Option nullOption;
+	nullOption.key = 0;
+	nullOption.value = nullptr;
+	m_options.push_back(nullOption);
+}
+
+
+
+
+void Lv2Options::initOption(LV2_URID key, uint32_t size, LV2_URID type,
+	const void *value,
+	LV2_Options_Context context, uint32_t subject)
```

Comment:
This function requires the caller to specify the type, its size, and a value of that type and size. This (a) is redundant information, which could be automatically determined, and (b) will result in undefined/undesirable behaviour if any of these arguments doesn't match the others. At the same time, I don't think the type should be determined from whatever variable the caller passes as the `value` parameter - this is an implementation detail and shouldn't be shared with LV2 plugins. Doing so would force the caller to use the exact type LV2 expects, and this requirement wouldn't be clear at the point the variable was declared. Thus I don't think overloading or templating is the solution here. I would go for a family of functions, which would enforce the correct type where the function is called:
```cpp
void initIntOption(LV2_URID key, std::int32_t value, LV2_Options_Context context, std::uint32_t subject);
void initFloatOption(LV2_URID key, float value, LV2_Options_Context context, std::uint32_t subject);
// etc...
```

Regarding option storage, I might consider using a `std::shared_ptr` here. While this has a little more overhead than a `std::unique_ptr`, it has the useful property that it can be passed polymorphically. This means you don't need to handle manual allocation and copying of the right amount of data, don't need a custom deleter, and it extends easily to types that have non-trivial destructors or for which `std::is_trivially_copyable_v` isn't true. E.g.:
```cpp
void Lv2Options::initOption(LV2_URID key, uint32_t size, LV2_URID type,
	std::shared_ptr<void> value,
	LV2_Options_Context context, uint32_t subject)
{
	// `value` can be used directly as storage
}
void initIntOption(LV2_URID key, std::int32_t value, LV2_Options_Context context, std::uint32_t subject)
{
	initOption(key, sizeof(value), Lv2UridCache::Id::atom_Int, std::make_shared<std::int32_t>(value), context, subject);
}
void initFloatOption(LV2_URID key, float value, LV2_Options_Context context, std::uint32_t subject)
{
	initOption(key, sizeof(value), Lv2UridCache::Id::atom_Float, std::make_shared<float>(value), context, subject);
}
// etc...
```

Reply:
Does that mean `std::shared_ptr<void>` can be used as a container for most types with proper deletion? If that's true, sounds cool.

Reply:
> I would go for a family of functions, which would enforce the correct type where the function is called

Can we make a template helper class which gives the atom URID for the given type, if it helps?
```cpp
// In class Lv2UridCache
template <typename T>
struct AtomUridForTypeHelper {};

template <>
struct AtomUridForTypeHelper<int> { static const Id value = Id::atom_Int; };

template <>
struct AtomUridForTypeHelper<float> { static const Id value = Id::atom_Float; };

template <typename T>
constexpr int AtomUridForType = AtomUridForTypeHelper<T>::value;

// In class Lv2Options
template <typename T>
void initOption(LV2_URID key, T value, LV2_Options_Context context, std::uint32_t subject)
{
	initOption(key, sizeof(value), Lv2UridCache::AtomUridForType<T>, std::make_shared<T>(value), context, subject);
}
```

Reply:
Considering the `sizeof(value)`... I thought about that, too, but in case you want to pass an array of `float`, the size calculation is wrong. There are currently no options that I know being passed as arrays, but there are even extension outside of the standard. Which is why I want to avoid `sizeof` in general (but `initFloatOption` makes clear that it's not a float array option).

Reply:
Same reason maybe for deriving the type...

Reply:
> in case you want to pass an array of `float`, the size calculation is wrong.

There are several ways to disallow passing problematic types, but that may be complicated.

Reply:
> Does that mean std::shared_ptr<void> can be used as a container for most types with proper deletion?

Yeah - it uses a neat trick internally. You have something like this (simplified, incomplete, hacked together in the early hours of the morning, not checked, probably wrong):
```cpp
class SharedPointerControlBlockBase
{
public:
    virtual ~SharedPointerControlBlockBase();
    std::size_t m_refcount;
};

template<typename T>
class SharedPointerControlBlockStorage : public SharedPointerControlBlockBase
{
public:
    template<typename... Args>
    SharedPointerControlBlockStorage(Args&&... args) : m_t(std::foward<Args>(args)...) { }
    T* getT() { return &m_t; }
private:
    T m_t;
};

template<typename T>
class shared_ptr
{
public:
    shared_ptr(T* value, SharedPointerControlBlockBase* control) : m_value{value}, m_control{control} { }
    template<typename U>
    shared_ptr(shared_ptr<U> &other) : m_value{other.m_value}, m_control{other.m_control} { ++m_control.m_refcount; }
    T* get() { return m_value; }
    ~shared_ptr() { if(--m_control.m_refcount == 0) { delete m_control; } }
private:
    T* m_value;
    SharedPointerControlBlockBase* m_controlBlock;
};

template<typename T, typename... Args>
shared_ptr<T> make_unique(Args&&... args)
{
    const auto control = new SharedPointerControlBlockStorage<T>(std::forward<Args>(args)...);
    return shared_ptr{control.getT(), control);
}
```
Each different type stored has its own type of control block, which deletes that type correctly. `shared_ptr` stores a pointer to the base control block type, which has a virtual destructor resulting in the deletion of the actual control block type. Once an appropriate control block is allocated, `shared_ptr`s with any compatible pointer type may use it. The stored pointer may be converted to a void pointer, but the actual object is still stored in the control block.

This is a useful pattern to know in C++, giving semantics similar to Rust's boxed traits. `shared_ptr` uses it to store "anything deletable", and `std::function` uses a similar trick to store "anything callable with the specified arguments". This was the approach I was going to suggest here, before I realised it would just be a reimplementation of half of `shared_ptr` with a lot more code.

Reply:
> Can we make a template helper class which gives the atom URID for the given type, if it helps?

I considered that, but you still have to manually associate each URID with each type, as with the functions. It also brings back the template I wanted to avoid:
```cpp
template <typename T>
void initOption(LV2_URID key, T value, LV2_Options_Context context, std::uint32_t subject);
```
This forwards the type `T`, which is some internal type used by the caller, all the way to an LV2 plugin. A generic API like this doesn't enforce the appropriate type for whichever option is being set. If a float option is required, for example, the caller is required to have an actual `float` variable to pass to this generic function. With a function that accepts `float` specifically, the caller is no longer restricted to using a `float` to store data.

For example, in LMMS we have `typedef uint32_t sample_rate_t;`, but LV2 wants the sample rate as a float (or at least that's what this PR does). Setting the option like this would be incorrect:
```cpp
m_options.initOption(Lv2UridCache::Id::param_sampleRate, Engine::mixer()->processingSampleRate());
```
This would set an integer option, not a floating point option. The type used by LMMS internally is taken, and exposed to the LV2 plugin. Yes, it's possible to add a cast to `float`, but I don't think that's a particularly obvious way to specify the option type. Also, such calls are susceptible to being broken easily, when the type of some internal value that used to be appropriate for an option is changed to something that isn't.

In short, this breaks encapsulation, and while it may be useful as an implementation detail for options, I don't think it makes a good API for them.

Reply:
Is this what you're saying? "Automatically guessing types is not good for APIs which may be used with implicit type conversion."

Reply:
Perhaps something more like "automatically guessing types is not good for APIs where using the correct type is important." With most APIs, either the API designer specifies parameter types explicitly (most cases out there), or accepts any reasonable type (generic APIs, such as standard algorithms, for example). If a range of types is accepted, but there's One True Type that should be passed, that sounds like a recipe for bugs to me.

| | Uses specific type | Uses any type |
|---|---|---|
| Accepts specific type | Good | Unnecessarily restrictive |
| Accepts any type | Will result in bugs | Good |

But I'm not an authority on this at all, so if you disagree, feel free to argue of course. (And if LV2 plugins are able to accept a variety of types for each option, my points are irrelevant anyway. It's not clear to me whether this is the case or not.)

Reply:
Sounds like a good guideline to me.

Reply:
Not sure why I didn't think of this ealier - the template approach is fine as long as the type parameter isn't deduced. Something like this would work, using the URID helper from before:
```cpp
template<typename Opt, typename Arg>
void initOption(LV2_URID key, Arg&& value, LV2_Options_Context context, std::uint32_t subject)
{
	initOption(key, sizeof(Opt), Lv2UridCache::AtomUridForType<Opt>,
		std::make_shared<Opt>(std::forward<Arg>(value)), context, subject);
}

Reply:
@DomClark the non-deduced example is interesting, but consider things like `char*`... Some users would just try `initOption<char>(...)`, and be surprised that only the first letter will be stored (since `sizeof(char) == 1`). Maybe this could be solved with an array variant, the same way as `new` and `new[]` do it?

Reply:
Yeah, I would overload it for array types. I would expect `initOption<char>` to initialise an option with a single `char`, the same way `initOption<int32_t>` would initialise an option with a single `int32_t`. For a string of chars, I would expect to use either `initOption<char*>` if I wanted to use a pointer to an existing string, or `initOption<char[]>` if I wanted the option to hold the whole string itself (both would just pass a `char*` to the LV2 plugin).

Is this required particularly soon, though? `std::shared_ptr` unfortunately only supports arrays since C++17.

Reply:
If you want to pass a whole array (not just a pointer to an array) of constexpr size, I would have rather expected `initOption<size, int32_t>`.

But it's indeed pure theory, there are no array options yet. I think I'll implement it like this now.

Reply:
@PhysSong Considering `AtomUridForTypeHelper`, I get `error: explicit specialization in non-namespace scope 'class Lv2UridCache'`. If I put it into a helper namespace outside, it works, but then, I can't use the `Id` class. I wonder how to properly solve this... I could move `Id` completely out pf `Lv2UridCache` :-1: or just use function overloading instead of template classes (`static constexpr Id atomUridForTypeHelper(int) { return Id::atom_Int; }`). Anyone a better solution?

Reply:
@DomClark What do we win (in practice) by using a `shared_ptr`? I think the sole advantage in this case is that we don't need a custom deleter? If that's the only thing, I wonder if it's worth using a `shared_ptr` in a place where the pointer is actually meant to never be shared (because that could make you think it is meant for being shared, plus the atomic counter overhead).

Reply:
I found that this works. However, we can fall back to https://github.com/LMMS/lmms/pull/5761#discussion_r523334501 if this increases complexity too much.
```cpp
class Class
{
public:
	enum class Id { _42 = 42 };
	template <typename T>
	struct sub;
	/*
	// error, see CWG 727
	template <>
	struct sub<int>
	{
		static constexpr Id value = Id::_42;
	};
	*/

	template <typename T>
	static constexpr Id sub_v = sub<T>::value;
};

template <> struct Class::sub<int>
{
	static constexpr Id value = Id::_42;
};
```

Reply:
@PhysSong Your proposal to resolve this at compile-time did not work, because the options must be inited with URIDs, not with `Lv2UridCache::Id`s. I wrote a work-around with function overloading, which uses the instance of type `Lv2UridCache` to resolve the Ids into URIDs and then pass them to the second `initOptions` in 85a3922.

Reply:
> If I put it into a helper namespace outside, it works, but then, I can't use the `Id` class.

`Id` is public, so you can just write `Lv2UridCache::Id`, right? Or am I missing something?

> the options must be inited with URIDs, not with `Lv2UridCache::Id`s.

The URID can be looked up in the URID cache in `initOption` at runtime, so determining an `Lv2UridCache::Id` at compile time should be alright, I think.

> What do we win (in practice) by using a `shared_ptr`? I think the sole advantage in this case is that we don't need a custom deleter?

No need for a custom deleter, and no need to manually allocate the right amount of space and copy the data in. Support for types that aren't trivially destructible or trivially copyable isn't relevant at the moment, and may well never be relevant since LV2 is a C API. But I like the peace of mind that it can't possibly go wrong, and it doesn't hurt to be flexible, in case we want to do something like attach a custom destructor to some data in the future.

> I wonder if it's worth using a shared_ptr in a place where the pointer is actually meant to never be shared (because that could make you think it is meant for being shared, plus the atomic counter overhead).

I don't think the overhead is significant (in an absolute sense, anyway - `shared_ptr<float>` probably uses several times more memory than `unique_ptr<float>`, but it's not like we're making thousands and thousands of these). Computational overhead should also be low - these things just sit untouched in a vector for most of their life, the reference count only touched at construction and destruction, when the allocator overhead is probably greater anyway. If profiling shows them to be a problem, of course we can rethink. As for being unintuitive, apparently this is an established pattern (try searching the internet for "shared_ptr void", although I admit I hadn't heard of it before), and a comment can always clear things up.

As for the function overloading, it's a reasonable work around. I feel it's a little less elegant/intuitive than PhysSong's approach, as it requires introducing values to determine a value that should be determinable using only types, but that's not really a problem.

If you do want to go back to using `unique_ptr`, I'd suggest adding `static_assert(std::is_trivially_copy_constructible<Opt>::value && std::is_trivially_destructible<Opt>::value);` to `initOption`, to avoid unpleasant surprises in the future.

Reply:
Per the `shared_ptr` vs `unique_ptr` - I just tried to rewrite the `shared_ptr` code to use `unique_ptr` instead. I failed, since `make_unique` does not support supplying a custom deleter. I like the idea of the `shared_ptr<void>` idiom, it looks more elegant after all to me, so I'll stick to this.

About the slowdown of `shared_ptr`, doesn't it use an atomic variable (while `unique_ptr` does not)? I once rewrote code to use atomics instead of normal variables, and had slowdowns of factor ~32. However, this here is just initialization code, outside of the realtime thread, so it should not matter.

Per looking up those `Lv2UridCache::Id` at compile time, I gave up, but you can try to write a patch if you like. Would be fun to see it work.

Reply:
> About the slowdown of `shared_ptr`, doesn't it use an atomic variable (while `unique_ptr` does not)? I once rewrote code to use atomics instead of normal variables, and had slowdowns of factor ~32.

`shared_ptr` only uses atomic operations for the reference count(s) by default. This code only ever moves, not copies, `shared_ptr`s, which won't touch the reference count. The only place where they would have an effect is at construction and destruction of the `shared_ptr`, but I would still put my money on this being less impactful than the memory allocation/deallocation.

> However, this here is just initialization code, outside of the realtime thread, so it should not matter.

Indeed.

> Per looking up those Lv2UridCache::Id at compile time, I gave up, but you can try to write a patch if you like. Would be fun to see it work.

I might have a go at it this weekend.

Reply:
This seems to work, mostly following PhysSong's comment above. (Only tested under MSVC, but should work everywhere, I think.)
```diff
diff --git a/include/Lv2Options.h b/include/Lv2Options.h
index 5bc3496b3..be1d6b207 100644
--- a/include/Lv2Options.h
+++ b/include/Lv2Options.h
@@ -69,9 +69,8 @@ public:
                        LV2_Options_Context context = LV2_OPTIONS_INSTANCE,
                        std::uint32_t subject = 0)
        {
-               using OptPtr = Opt*;
                const Lv2UridCache& cache = Engine::getLv2Manager()->uridCache();
-               initOption(cache[key], sizeof(Opt), cache.uridForType(OptPtr()),
+               initOption(cache[key], sizeof(Opt), cache[Lv2UridCache::IdForType_v<Opt>],
                        std::make_shared<Opt>(std::forward<Arg>(value)), context, subject);
        }
        //! Fill m_options and m_optionPointers with all options
diff --git a/include/Lv2UridCache.h b/include/Lv2UridCache.h
index 21d1a0e6d..8f32a4163 100644
--- a/include/Lv2UridCache.h
+++ b/include/Lv2UridCache.h
@@ -50,18 +50,24 @@ public:
                // exception to alphabetic ordering - keep at the end:
                size
        };
+
+       template<typename T>
+       struct IdForType;
+
+       template<typename T>
+       static constexpr auto IdForType_v = IdForType<T>::value;
+
        //! Return URID for a cache ID
        uint32_t operator[](Id id) const;

-       // get URID by type
-       uint32_t uridForType(const int32_t*) const { return operator[](Id::atom_Int); }
-       uint32_t uridForType(const float*) const { return operator[](Id::atom_Float); }
-
        Lv2UridCache(class UridMap& mapper);

 private:
        uint32_t m_cache[static_cast<int>(Id::size)];
 };

+template<> struct Lv2UridCache::IdForType<float> { static constexpr auto value = Id::atom_Float; };
+template<> struct Lv2UridCache::IdForType<std::int32_t> { static constexpr auto value = Id::atom_Int; };
+
 #endif // LMMS_HAVE_LV2
 #endif // LV2URIDCACHE_H
```

Reply:
Many thanks both of you! I added your patch and reviewed it, so now the whole code is reviewed.